### ISSUE TITLE
Investigating the Bullet gem to help improve performance for DB queries

### ISSUE THREAD

**Issue by julianguyen** [ifmeorg/ifme#1160]
# Description 

@jm96441n made mention of this [gem](https://github.com/flyerhzm/bullet) in a recent [pull request](https://github.com/ifmeorg/ifme/pull/1143) I made refactoring model and controller code related to Comments (thank you!)

It would be great to investigate this gem further and use it to improve performance for DB queries.

---

Please assign yourself (via the Assignees dropdown), if you do want to work on this issue. Can't find yourself? You need to [join our organization](https://github.com/ifmeorg/ifme/wiki/Join-Our-Slack).

Check out our [Picking Up Issues](https://github.com/ifmeorg/ifme/wiki/Picking-Up-Issues) guide if you haven't already!

**Comment by gabrielpedepera** [comment]
Hi @julianguyen, I would like to work on this, may I?

**Comment by muratbsts** [comment]
@julianguyen @gabrielpedepera if you'd not like to work on this, I'd really like.

**Comment by marcellalcantara** [comment]
Just to keep everyone updated, @gabrielpedepera  and I installed the [Bullet](https://github.com/flyerhzm/bullet) gem and executed in the application.

Our first strategy was to run the specs, so we already have a `bullet.log` with some warnings to improve.

We will send the PR with the first improvements soon.

**Comment by subwindow** [comment]
If no one is still working on this I'd like to take it.

**Comment by julianguyen** [comment]
@subwindow Not sure if @marcelkalveram and @gabrielpedepera are still working on it. But feel free to take a stab at it! :)

**Comment by marcellalcantara** [comment]
Hello. 
Sorry for the delay in answering. We have already installed the tool but have not yet had time to work on it. As soon as we start working, we’ll let you know. Hugs

**Comment by cartothemax** [comment]
Looks like this issue might be good to be closed? 😄 Nice work @gabrielpedepera and @marcellalcantara

**Comment by julianguyen** [comment]
It would appear so @cartothemax! I guess the only thing left to do is actually use the tool to improve query performance. Thoughts @gabrielpedepera and @marcellalcantara? :)

**Comment by andrew-schutt** [comment]
I will spend some time investigating the output from `bullet.log` locally and see if we can't get some actionable issues spun up similar to the `controller` -> `request` spec refactoring that is going on right now. Similar to this collection of issues https://github.com/ifmeorg/ifme/issues/1815

**Comment by julianguyen** [comment]
Awesome thanks!!

**Comment by andrew-schutt** [comment]
Been spending a little bit of time reading up on the [bullet gem](https://github.com/flyerhzm/bullet). Couple good resources that I found:
- http://railscasts.com/episodes/372-bullet
- https://www.123ish.com/en/entries/2235-optimize-your-rails-code-with-bullet-gem-and-speed-up-your-site-
- https://semaphoreci.com/blog/2017/08/09/faster-rails-eliminating-n-plus-one-queries.html
- https://medium.com/@codenode/10-tips-for-eager-loading-to-avoid-n-1-queries-in-rails-2bad54456a3f

The Rails Cast episode discusses what "success" looks like and also warns of just blindly using `includes`. This led me to wonder if introducing performance tests around spots to ensure that we are improving performance. There is of course a [Rails Guide on performance testing](https://guides.rubyonrails.org/v3.2.13/performance_testing.html). 

Our `bullet.log` output in `log/` has quite a few warnings on my machine. That being said all of them fall into two categories of either `AVOID` or `USE` eager loading. Additionally, they are being generated around `groups` queries specifically.

I think to address the warnings I will create some basic performance testing around the `/groups` CRUD operators, figure out our benchmarks, implement the suggested changes from `bullet`, and see where we are after that.

**Comment by julianguyen** [comment]
This is awesome @andrew-schutt, thanks for looking into this and sharing resources! That sounds like a solid plan to me!

**Comment by andrew-schutt** [comment]
Alright here's a report back on my findings

For testing the performance of the `groups_controller` I ended up using the [rack-mini-profiler gem](https://github.com/MiniProfiler/rack-mini-profiler) This includes a small timing for loading the page in the upper left-hand corner.

![Screen Shot 2020-10-06 at 8 54 08 AM](https://user-images.githubusercontent.com/413224/95256669-ba0aa900-07e8-11eb-9b7c-fd27f60f0512.png)

This will only be present in the dev and test environments. After installing and setting up the gem I tried to get a feel for what the current response time was for `/groups`. Then per the `log/bullet.log`, I tired out removing the `.includes(:group_members)` call from the `GroupsController#index` action. To be honest, I saw _maybe_ incremental improvements, but nothing drastic.

That ended up being really the only actionable warning though. The other warnings were recommendations to include `:leaders` on queries, but this isn't really possible as it's an aliased `has_many` on the `GroupMembers` model 🤷‍♂️ 

The good news is that the project is pretty clean! It's nice to have the bullet gem in the project and will be a useful tool 😄
